tt_metal/tt_metal.cpp

```diff
@@ -364,14 +364,27 @@ void CloseDevices(std::map<chip_id_t, Device *> devices) {
             for (uint32_t ts = t.size() - 1; ts > 0; ts--) {
                 if (devices.find(t[ts]) != devices.end()) {
                     devices[t[ts]]->close();
+                    // When a device is closed, its worker thread is joined. Stop tracking this
+                    // worker thread.
+                    tt::DevicePool::instance().delete_worker_thread_for_device(devices[t[ts]]);
                 }
             }
         }
         //finally close the mmio device
         dev->close();
+        tt::DevicePool::instance().delete_worker_thread_for_device(dev);
     }
 }
 
+bool InApplicationThread() {
+    bool in_application_thread = true;
+    if (tt::DevicePool::is_instantiated()) {
+        auto worker_thread_ids = tt::DevicePool::instance().get_worker_thread_ids();
+        in_application_thread = worker_thread_ids.find(std::hash<std::thread::id>{}(std::this_thread::get_id())) == worker_thread_ids.end();
```

Comment:
Will this work w/o specification of std::hash<..>?
```suggestion
        auto worker_thread_ids = tt::DevicePool::instance().get_worker_thread_ids();
        const auto this_thread_id = std::this_thread::get_id();
        in_application_thread = worker_thread_ids.find(this_thread_id) == worker_thread_ids.end();
```

Reply:
Good point, I've updated these functions to work with raw thread ids, instead of hashes